Fix use of account key() as seed

mcintyre94 opened the pull request:
The code we were generating was `user . borrow () . __account__ . key () . as_ref ()`. But `borrow()` didn't exist on the boxed accounts that we end up with in the accounts struct.

Instead we can just use `user.key().as_ref()` directly. 

Closes #62

ameliatastic commented:
Does this still allow `.key()` to be used in non-seed contexts?

mcintyre94 commented:
Oops looks like this needs a bit more work, I added this instruction:

```py
@instruction
def hello(payer: Signer, user: User, another: Empty[Another]):
    a = another.init(payer, seeds=[user.key()])
    k = user.key()
    print(k)
```

And got this error:
```
anchor build -p seahorse_seeds failed:
   Compiling seahorse_seeds v0.1.0 (/Users/callum/scratch/seahorse_seeds/programs/seahorse_seeds)
error[E0599]: no method named `key` found for struct `seahorse_util::Mutable<LoadedUser<'info, '_>>` in the current scope
  --> programs/seahorse_seeds/src/dot/program.rs:85:22
   |
85 |     let mut k = user.key();
   |                      ^^^ method not found in `seahorse_util::Mutable<LoadedUser<'info, '_>>`
   |
  ::: programs/seahorse_seeds/src/lib.rs:25:5
   |
25 |     pub struct Mutable<T>(Rc<RefCell<T>>);
   |     -------------------------------------- method `key` not found for this
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `key`, perhaps you need to implement it:
           candidate #1: `anchor_lang::Key`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `seahorse_seeds` due to previous error
```

Marked this as draft, I'll take another look at this

ameliatastic commented:
Just adding some thoughts, I think the very general problem here is that the AST (in the _build_ step) is built bottom-up. This means that expressions like `[account.key()]` are completely built without any regard for the context that they're in, which might be a "normal" expression (`keys = [account.key()]`) or a "special" expression (`my_account.init(seeds = [account.key()], ...`).

So a full solution to this might require some real compiler rework. Maybe just pass down a context object to enable/disable certain behaviors and use that info to build up the AST?

mcintyre94 commented:
I think it would need to be something like that, not sure exactly how that'd look though! IIRC didn't 0.1 have some special handling for seeds, where they'd be rendered specifically? I guess a generalised and more flexible version of that would be nice!

mcintyre94 commented:
@ameliatastic I've refactored this, could you review please?

- I've removed the special case expression rendering for `key()` from the check stage, it's just returned as the existing expression there now
- I've added a special case in the build stage, which renders it to `account.borrow().__account__.key()` when used in a non-seed. Other attributes (eg class fields), and `key()` in a seed are just compiled as normal
- Added a test case. It's in a new directory `tests/test-cases` which is intended for stuff that's worth having tests for but doesn't make sense to put in an example. Behaves exactly the same as examples in the CI, they get compiled and there's an error if there's any mismatch. I'm keen not to limit our tests to things that can double as useful documentation, which examples was forcing us into a bit.

ameliatastic commented:
Hey, really sorry I missed this, didn't realize you had brought this PR out of draft. Looking at it now!

ameliatastic commented on src/core/compile/check/mod.rs:
```diff
@@ -791,14 +791,7 @@ impl<'a> Context<'a> {
-                                            #account.borrow().__account__.key()
-                                        });
-
+                                    Transformation::new(|expr| { 
```
I think that the special case you added to the build stage can be implemented here - you can use Transformation::new_with_context to create a transformation that can use the context stack.

Otherwise, if I'm wrong, you can remove this transformation entirely since it wouldn't do anything (just make `Pubkey` the return type of the `key` function instead of `Ty::Transformed(...)`).

mcintyre94 replied:
Thanks, that makes sense! I've tried moving the same logic into a `new_with_context` block in check:

```rs
Transformation::new_with_context(|mut expr, context_stack| {
    if !context_stack.has(&ExprContext::Seed) {
        // If not in a seed, rewrite account.key() to account.borrow().__account__.key()
        let value = ExpressionObj::BorrowImmut(expr.obj.into());
        let value = ExpressionObj::Attribute {
            value: value.into(),
            name: String::from("__account__"),
        };
        let value = ExpressionObj::Attribute {
            value: value.into(),
            name: "key".to_string(),
        };
        expr.obj = value;
    }
    Ok(Transformed::Expression(expr))
}, None)
```

But for some reason it doesn't quite work the same, and the output I'm getting (for non seed cases) looks like:

```rs
calculator.key().borrow().__account__.key
```

So when I do `expr.obj()` it's getting `calculator.key()` where in the build stage it was getting `calculator` as the base I built on

And when I do the `ExpressionObj::Attribute` at the end, it's writing out `.key` instead of `.key()` as it did in the current code

Do you know what would cause those differences/how I could fix them in the check stage?

mcintyre94 replied:
Okay cool I think I've got it working, and now I understand what this stuff is doing!

```rs
let account = match1!(expr.obj, ExpressionObj::Call { function, .. } => *function);
let account = match1!(account.obj, ExpressionObj::Attribute { value, .. } => *value);
```

Added those back to get to the `account` from `account.key()` and it's looking good!

ameliatastic reviewed (approved):
Everything lgtm! Test compiled myself and the example worked, I'll merge this.
